linera-views/src/backends/lru_caching.rs

```diff
@@ -48,138 +308,428 @@ static NUM_CACHE_SUCCESS: LazyLock<IntCounterVec> = LazyLock::new(|| {
     const MAX_KEY_SIZE: usize = K::MAX_KEY_SIZE;
-    type Keys = K::Keys;
-    type KeyValues = K::KeyValues;
+    type Keys = Vec<Vec<u8>>;
+    type KeyValues = Vec<(Vec<u8>, Vec<u8>)>;
 
     fn max_stream_queries(&self) -> usize {
         self.store.max_stream_queries()
     }
 
     async fn read_value_bytes(&self, key: &[u8]) -> Result<Option<Vec<u8>>, Self::Error> {
-        let Some(lru_read_values) = &self.lru_read_values else {
+        let Some(cache) = &self.cache else {
             return self.store.read_value_bytes(key).await;
         };
-        // First inquiring in the read_value_bytes LRU
+        // First inquiring in the read_value_bytes cache
         {
-            let lru_read_values_container = lru_read_values.lock().unwrap();
-            if let Some(value) = lru_read_values_container.query(key) {
+            let mut cache = cache.lock().unwrap();
+            if let Some(value) = cache.query_read_value(key) {
                 #[cfg(with_metrics)]
-                NUM_CACHE_SUCCESS.with_label_values(&[]).inc();
-                return Ok(value.clone());
+                NUM_CACHE_VALUE_SUCCESS.with_label_values(&[]).inc();
+                return Ok(value);
             }
         }
         #[cfg(with_metrics)]
-        NUM_CACHE_FAULT.with_label_values(&[]).inc();
+        NUM_CACHE_VALUE_FAULT.with_label_values(&[]).inc();
         let value = self.store.read_value_bytes(key).await?;
-        let mut lru_read_values = lru_read_values.lock().unwrap();
-        lru_read_values.insert(key.to_vec(), value.clone());
+        let mut cache = cache.lock().unwrap();
+        cache.insert_read_value(key.to_vec(), &value);
         Ok(value)
     }
 
     async fn contains_key(&self, key: &[u8]) -> Result<bool, Self::Error> {
-        if let Some(values) = &self.lru_read_values {
-            let values = values.lock().unwrap();
-            if let Some(value) = values.query(key) {
-                return Ok(value.is_some());
+        let Some(cache) = &self.cache else {
+            return self.store.contains_key(key).await;
+        };
+        {
+            let mut cache = cache.lock().unwrap();
+            if let Some(result) = cache.query_contains_key(key) {
+                #[cfg(with_metrics)]
+                NUM_CACHE_VALUE_SUCCESS.with_label_values(&[]).inc();
+                return Ok(result);
             }
         }
-        self.store.contains_key(key).await
+        #[cfg(with_metrics)]
+        NUM_CACHE_VALUE_FAULT.with_label_values(&[]).inc();
+        let result = self.store.contains_key(key).await?;
+        let mut cache = cache.lock().unwrap();
+        cache.insert_contains_key(key.to_vec(), result);
+        Ok(result)
     }
 
     async fn contains_keys(&self, keys: Vec<Vec<u8>>) -> Result<Vec<bool>, Self::Error> {
-        let Some(values) = &self.lru_read_values else {
+        let Some(cache) = &self.cache else {
             return self.store.contains_keys(keys).await;
```

Comment:
If we have cache we check it and return immediately. What about the case where we have a cache, but not all keys are there. I'd expect that we check underlying storage and cache the newly-requested keys.